### PR TITLE
ARROW-7888: [Python] Update pyarrow.jvm to support jpype 0.7+

### PR DESCRIPTION
**Pull request by mariusvniekerk**
Not quite sure where the test suite for this is disabled / enabled?

### CODE REVIEW

**Comment by pitrou** [comment]
@mariusvniekerk The test suite is in `test_jvm.py`.

**Comment by pitrou** [comment]
It's probably not enabled in CI, if that's what you're asking :-) It should be easy to enable, though.

**Comment by pitrou** [comment]
I've tried this PR and it does fix some test errors but there are still a couple of them:
https://gist.github.com/pitrou/e4859515a1cae233aeec7aa655192132

**Comment by mariusvniekerk** [comment]
@pitrou Okay, this might be enough.  The ambiguity error is due to jpype not really seeing much difference between int and char.  It looks like we can cast to the correct int type to get things to behave as expected.

**Comment by wesm** [comment]
It could be less onerous to test this in a Crossbow / docker-compose job, not sure if there is a JIRA about this already

**Comment by mariusvniekerk** [comment]
Not quite sure how to set that up @wesm

**Comment by pitrou** [comment]
I've pushed further changes. @mariusvniekerk do you want to give them a quick look?

**Comment by wesm** [comment]
@mariusvniekerk my comment was mainly for @kszucs or others who are more the experts in the docker-compose configurations =)

**Comment by kszucs** [comment]
Antoine has created a jira to track this https://issues.apache.org/jira/browse/ARROW-7930

**Comment by kszucs** [comment]
@ursabot crossbow submit test-conda-python-3.8-jpype

File: python/pyarrow/public-api.pxi [hunk 1]
```diff
@@ -130,7 +130,8 @@ cdef object pyarrow_wrap_metadata(
     return result
 
 
-cdef shared_ptr[CKeyValueMetadata] pyarrow_unwrap_metadata(object meta):
+cdef shared_ptr[CKeyValueMetadata] pyarrow_unwrap_metadata(object meta) \
```
Comment (xhochy) [inline python/pyarrow/public-api.pxi]: Does this also introduce a change in the generated C++ header? Do we need to update https://github.com/apache/arrow/blob/master/cpp/src/arrow/python/pyarrow_api.h ?

Comment (pitrou) [reply python/pyarrow/public-api.pxi]: Not really, but updating anyway.

**Comment by mariusvniekerk** [comment]
Thanks @pitrou

**Review by xhochy — APPROVED** [review]
LGTM

**Comment by pitrou** [comment]
Rebased.

**Comment by pitrou** [comment]
GHA failures are spurious.

**Comment by pitrou** [comment]
This was merged in 4aded1dd9804e4ca12c63fcd32fa2d7e2f5d1922. Apparently the Github UI didn't notice.

**Pull request closed without merging**